state/applicationoffers_test.go

```diff
@@ -566,6 +653,27 @@ func (s *applicationOffersSuite) TestRemoveOffersWithConnectionsForce(c *gc.C) {
 	s.assertInScope(c, mysqlru, true)
 }
 
+func (s *applicationOffersSuite) TestRemovingApplicationFailsRace(c *gc.C) {
+	s.createDefaultOffer(c)
+	wp := s.AddTestingApplication(c, "local-wordpress", s.AddTestingCharm(c, "wordpress"))
+
+	bumpRelCount := func() {
+		eps, err := s.State.InferEndpoints(wp.Name(), s.mysql.Name())
+		_, err = s.State.AddRelation(eps...)
+		c.Assert(err, jc.ErrorIsNil)
+	}
+
+	bumpRevno := func() {
```

Comment:
The key thing we are guarding against is that the addition and removal of a relation meaning the rel count stays the same. So we only need bumpRelCount I think? As that will inc the rel count and hence the txn rev no? Why do we need this bumpRevNo func?

Reply:
It's just a little bit simpler. I can create three relations, but each needs a unique name, but that muddies up the test.

An alternative is to delete the relation each time--this is actually what I started with--but it tends to generate a near-impossible-to-debug error message:

These test hooks

```go
addRelation := func() {
	_, err = s.State.AddRelation(eps...)
	c.Assert(err, jc.ErrorIsNil)
}

rmRelations := func() {
	rels, err := s.State.AllRelations()
	c.Assert(err, jc.ErrorIsNil)

	for _, rel := range rels {
		rel, err := s.State.Relation(rel.Id())
		c.Assert(err, jc.ErrorIsNil)
		err = rel.Destroy()
		c.Assert(err, jc.ErrorIsNil)
	}
}

hooks := []txn.TestHook{
	{Before: addRelation, After: rmRelations},
	{Before: addRelation, After: rmRelations},
	{Before: addRelation, After: rmRelations},
}

state.SetTestHooks(c, s.State, hooks...).Check()
```
generates this pretty generic error message
```
c.Assert(remaining, gc.HasLen, 0)
... obtained []txn.TestHook = []txn.TestHook{txn.TestHook{Before:(func())(0x2f78a10), After:(func())(0x2f78bb0)}, txn.TestHook{Before:(func())(0x2f78a10), After:(func())(0x2f78bb0)}, txn.TestHook{Before:(func())(0x2f78a10), After:(func())(0x2f78bb0)}}
... n int = 0
```
I've tried running those two functions outside of the hooks, and they appear to be perfectly fine. Adding refreshes etc doesn't help.

The current test works and is simple. It's a shame that we will need to change the test if we change the schema to include a remote relation count. But that's a relatively small additional step for what would be a fairly extensive change.

Reply:
Figured it out with assistance from @babbageclunk. I left off the requred `defer` to `state.SetTestHooks().Check()`.